src/main/java/ru/rt/restream/reindexer/convert/util/ConversionUtils.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2020 Restream
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ru.rt.restream.reindexer.convert.util;
+
+import org.apache.commons.lang3.reflect.TypeUtils;
+import ru.rt.restream.reindexer.convert.FieldConverter;
+import ru.rt.restream.reindexer.util.Pair;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * For internal use only, as this contract is likely to change.
+ */
+public final class ConversionUtils {
+
+    private static final Map<Pair<Class<?>, String>, ResolvableType> FIELD_TYPE_CACHE = new ConcurrentHashMap<>();
+
+    /**
+     * Resolves source and target types based on {@code X} and {@code Y} parameters from {@link FieldConverter} interface implementation.
+     * If {@code FieldConverter} source or target type is an array, the type is determined by {@code Class#getComponentType},
+     * in case of a generic type, only {@code java.util.Collection} assignable types are considered, nested generic types
+     * (i.e. {@code List<List<String>>}) are not supported, such types will not be recognized and IllegalArgumentException will be thrown.
+     * @param converterClass the {@link FieldConverter} interface implementation class to use
+     * @return the {@link Pair} of source and target {@link ResolvableType}s to use
+     */
+    public static Pair<ResolvableType, ResolvableType> resolveConverterTypePair(Class<?> converterClass) {
```

Comment:
It seems the code can be made simpler if you use something like `Class<? extends FiledConverterClass<?,?>>` . Or implement the `getConvertiblePair()` method in the `FiledConverter` interface, and the method will not be needed at all.

Reply:
I agree this method can be omitted.